File: src/Library/ObjectModel/TimeZoneSetupStep.vb
Public Class TimeZoneSetupStep
    Inherits Pixa.Soundbridge.Library.SetupStep

    Friend Sub New(ByVal client As ISoundbridgeClient)
        MyBase.New(client)
    End Sub

    Protected Overrides ReadOnly Property Name() As String
        Get
            Return "Time Zone"
        End Get
    End Property

    Public Overrides Function GetSelectionList() As String()
        Return Client.ListTimeZones
    End Function

    Public Overrides Sub MakeSelection(ByVal value As String)
        Dim timeZones As String() = Client.ListTimeZones
        Dim index As Integer = Array.IndexOf(timeZones, value)

        If index = -1 Then Throw New Exception(String.Format("Couldn't find Time Zone '{0}' in the list from the Soundbridge", value))

        Dim r As String = Client.SetTimeZone(index)

        If r = "OK" Then Exit Sub
        If r = "ParameterError" Then Throw New ArgumentException("Error setting the Time Zone", "value")
        If r = "GenericError" Then Throw New Exception("Soundbridge returned GenericError")
    End Sub
End Class

File: src/Pixa.Soundbridge/TimeZoneSetupStep.cs
using Pixa.Soundbridge.Client;
using System;

namespace Pixa.Soundbridge
{
    public class TimeZoneSetupStep : SetupStep
    {
        internal TimeZoneSetupStep(ISoundbridgeClient client) : base(client)
        {
        }

        protected override string Name
        {
            get
            {
                return "Time Zone";
            }
        }

        public override string[] GetSelectionList()
        {
            return Client.ListTimeZones();
        }

        public override void MakeSelection(string value)
        {
            var timeZones = Client.ListTimeZones();
            int index = Array.IndexOf(timeZones, value);
            if (index == -1)
                throw new Exception(string.Format("Couldn't find Time Zone '{0}' in the list from the Soundbridge", value));
            string r = Client.SetTimeZone(index);
            if (r == "OK")
                return;
            if (r == "ParameterError")
                throw new ArgumentException("Error setting the Time Zone", "value");
            if (r == "GenericError")
                throw new Exception("Soundbridge returned GenericError");
        }
    }
}